src/typecheck.rs

```diff
@@ -1,42 +1,109 @@
-use crate::*;
-use indexmap::map::Entry as IEntry;
+use std::ops::AddAssign;
+
+use crate::{
+    ast::desugar::flatten_actions,
+    constraint::{all_equal_constraints, Assignment},
+    *,
+};
+use hashbrown::HashMap;
 use typechecking::TypeError;
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Default)]
+pub struct Actions(pub(crate) Vec<NormAction>);
+impl Actions {
+    fn subst(&mut self, subst: &HashMap<Symbol, AtomTerm>) {
+        let actions = subst.iter().map(|(symbol, atom_term)| match atom_term {
+            AtomTerm::Var(v) => NormAction::LetVar(*symbol, *v),
+            AtomTerm::Literal(lit) => NormAction::LetLit(*symbol, lit.clone()),
+            AtomTerm::Global(v) => NormAction::LetVar(*symbol, *v),
+        });
+        let existing_actions = std::mem::take(&mut self.0);
+        self.0 = actions.chain(existing_actions).collect();
+    }
+}
+
+// TODO: implement custom debug
+#[derive(Debug, Clone)]
+pub struct UnresolvedCoreRule {
+    pub body: Query<Symbol>,
+    pub head: Actions,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct ResolvedCoreRule {
+    pub body: Query<ResolvedSymbol>,
```

Comment:
Since you are doing a bunch of refactoring, it would be awesome if the resolved query had type annotations. But that's future work